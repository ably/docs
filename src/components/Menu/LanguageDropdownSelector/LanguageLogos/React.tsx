import React from 'react';

// The best one.
export const ReactJS = () => (
  <svg width="24" height="24" viewBox="-11.5 -10.23174 23 20.46348" fill="none" xmlns="http://www.w3.org/2000/svg">
    <circle cx="0" cy="0" r="2.05" fill="#61dafb" />
    <g stroke="#61dafb" strokeWidth="1" fill="none">
      <ellipse rx="11" ry="4.2" />
      <ellipse rx="11" ry="4.2" transform="rotate(60)" />
      <ellipse rx="11" ry="4.2" transform="rotate(120)" />
    </g>
  </svg>
);
