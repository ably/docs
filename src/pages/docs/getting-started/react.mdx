---
title: "Getting started: Pub/Sub with React"
meta_description: "A getting started guide for Ably Pub/Sub React that steps through some of the key features using React and Vite."
meta_keywords: "Ably, realtime, quickstart, getting started, basics, Pub/Sub, React, Vite"
languages:
  - react
---

This guide will get you started with Ably Pub/Sub in a new React application built with Vite.

It will take you through the following steps:

* Create a client and establish a realtime connection to Ably.
* Attach to a channel and subscribe to its messages.
* Publish a message to the channel for your client to receive.
* Join and subscribe to the presence set of the channel.
* Retrieve the messages you sent in the guide from history.
* Close a connection to Ably when it is no longer needed.

## Prerequisites <a id="prerequisites"/>

1. [Sign up](https://ably.com/signup) for an Ably account.
2. Create a new app, and create your first API key.
3. Your API key will need the `publish`, `subscribe`, `presence` and `history` capabilities.
4. Install the Ably CLI:

<Code>
```shell
npm install -g @ably/cli
```
</Code>

5. Run the following to log in to your Ably account and set the default app and API key:

<Code>
```shell
ably login

ably apps switch
ably auth keys switch
```
</Code>

<Aside data-type='note'>
The code examples in this guide include a demo API key. If you wish to interact with the Ably CLI and view outputs within your Ably account, ensure that you replace them with your own API key.
</Aside>

### Create a React project <a id="prerequisites-create-project"/>

Create a new React + TypeScript project using [Vite](https://vitejs.dev/guide/#scaffolding-your-first-vite-project). Then, navigate to the project folder and install the dependencies:

<Code>
```shell
npm create vite@latest ably-pubsub-react -- --template react-ts

│
◇  Scaffolding project in /ably-pubsub-react...
│
└  Done. Now run:

  cd ably-pubsub-react
  npm install
```
</Code>

You should see a directory structure similar to this:

<code>
  ├── index.html
  ├── package.json
  ├── public
  ├── src
  │   ├── assets
  │   ├── App.css
  │   ├── App.tsx
  │   ├── index.css
  │   ├── main.tsx
  │   └── vite-env.d.ts
  ├── tsconfig.app.json
  ├── tsconfig.json
  ├── tsconfig.node.json
  └── vite.config.ts
</code>

You will also need to setup [Tailwind CSS](https://tailwindcss.com/docs/installation/using-vite) for styling the application.
First, install the required Tailwind CSS packages:

<Code>
```shell
npm install tailwindcss @tailwindcss/vite
```
</Code>

Next, update `vite.config.ts` file to include the Tailwind CSS plugin:

<Code>
```react
// vite.config.ts
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import tailwindcss from '@tailwindcss/vite';

// https://vite.dev/config/
export default defineConfig({
  plugins: [react(), tailwindcss()],
});
```
</Code>

Finally, import Tailwind CSS in the `src/index.css` file and remove all other existing CSS styles:

<Code>
```css
/* src/index.css */
@import 'tailwindcss';
```
</Code>

And replace the contents of `src/App.tsx` with the following:

<Code>
```react
// src/App.tsx
function App() {
  return (
    <div className='flex flex-col w-[900px] h-full border-1 border-blue-500 rounded-lg overflow-hidden mx-auto font-sans'>
      <div className='flex flex-row w-full rounded-lg overflow-hidden mx-auto font-sans'>
        <div className='flex-1 bg-gray-100 text-center p-4'>
          <h2 className='text-lg font-semibold text-blue-500'>
            Ably Pub/Sub React
          </h2>
        </div>
      </div>
    </div>
  );
}

export default App;
```
</Code>

### Install Ably Pub/Sub JavaScript SDK <a id="prerequisites-ably-pubsub"/>

Install the Ably Pub/Sub JavaScript SDK in your React project:

<Code>
```shell
npm install ably
```
</Code>

### Set up AblyProvider <a id="prerequisites-setup-ably-provider"/>

The Ably Pub/Sub SDK provides React hooks and context providers that make it easier to use Pub/Sub features in your React components.

The `AblyProvider` component should be used at the top level of your application, typically in `main.tsx`. It provides access to the Ably Realtime client for all child components that use Ably Pub/Sub React hooks.

<Aside data-type='note'>
The `AblyProvider` is required when using the `useAbly()` and `useConnectionStateListener()` hooks, and the `ChannelProvider` exposed by the Ably Pub/Sub SDK.
</Aside>

Replace the contents of your `src/main.tsx` file with the following code to set up the `AblyProvider`:

<Code>
```react
// src/main.tsx
import * as Ably from 'ably';
import { AblyProvider } from 'ably/react';
import { StrictMode } from 'react';
import { createRoot } from 'react-dom/client';
import App from './App.tsx';
import './index.css';

// Create your Ably Realtime client
const realtimeClient = new Ably.Realtime({
  key: '{{API_KEY}}',
  clientId: 'my-first-client',
});

createRoot(document.getElementById('root')!).render(
  <StrictMode>
    <AblyProvider client={realtimeClient}>
      <App />
    </AblyProvider>
  </StrictMode>
);
```
</Code>

<aside data-type='note'>
Keep the Realtime client initialization outside of any React component to prevent it from being recreated on re-renders, which could result in reaching your Ably connection limit.
</aside>

## Step 1: Connect to Ably <a id="step-1"/>

Clients establish a connection with Ably when they instantiate an SDK instance. This enables them to send and receive messages in realtime across channels.

Open up the [dev console](https://ably.com/accounts/any/apps/any/console) of your first app before instantiating your client so that you can see what happens.

In the [Set up AblyProvider](#prerequisites-setup-ably-provider) section, you added the following code to create an Ably Realtime client:

<Code>
```react
const realtimeClient = new Ably.Realtime({
  key: '{{API_KEY}}',
  clientId: 'my-first-client',
});
```
</Code>

This code creates a new Realtime client instance, establishing a connection to Ably when your application starts. At the minimum you need to provide an authentication mechanism. While using an API key is fine for the purposes of this guide, you should use [][token authentication](/docs/auth/token) in production environments. A `clientId` ensures the client is identified, which is required to use certain features, such as presence.

To monitor the Ably connection state within your application, create a component that uses the `useConnectionStateListener()` hook provided by the Ably Pub/Sub SDK. This hook must be nested inside an `AblyProvider`, so the component must be placed within the `AblyProvider` in your application.

In your project, create a new file `src/ConnectionState.tsx` with the following content:

<Code>
```react
// src/ConnectionState.tsx

// React hooks are exported from the 'ably/react' path of the 'ably' package.
import { useAbly, useConnectionStateListener } from 'ably/react';
import { useState } from 'react';

export function ConnectionState() {
  // This component displays the current connection state

  // The useAbly hook returns the Ably Realtime client instance provided by the AblyProvider
  const ably = useAbly();
  const [connectionState, setConnectionState] = useState(ably.connection.state);

  // useConnectionStateListener hook listens for changes in connection state
  useConnectionStateListener((stateChange) => {
    setConnectionState(stateChange.current);
  });

  return (
    <div className='mt-4 text-center h-full'>
      <p>Connection: {connectionState}!</p>
    </div>
  );
}
```
</Code>

Then, update your `App` component in the `src/App.tsx` file to include the `ConnectionState` component:

<Code>
```react
// src/App.tsx

// Import your newly created component
import { ConnectionState } from './ConnectionState';

function App() {
  return (
    <div className='flex flex-col w-[900px] h-full border-1 border-blue-500 rounded-lg overflow-hidden mx-auto font-sans'>
      <div className='flex flex-row w-full rounded-lg overflow-hidden mx-auto font-sans'>
        <div className='flex-1 bg-gray-100 text-center p-4'>
          <h2 className='text-lg font-semibold text-blue-500'>
            Ably Pub/Sub React
          </h2>
          {/* Add ConnectionState here */}
          <ConnectionState />
        </div>
      </div>
    </div>
  );
}

export default App;
```
</Code>

Now run your application by starting the development server:

<Code>
```shell
npm run dev
```
</Code>

Open the URL shown in the terminal (typically [http://localhost:5173/](http://localhost:5173/)).

You should see the connection state displayed in your UI (e.g., `Connection: connected!`). You can also inspect connection events in the [dev console](https://ably.com/accounts/any/apps/any/console) of your app.

## Step 2: Subscribe to a channel and publish a message <a id="step-2"/>

Messages contain the data that a client is communicating, such as a short 'hello' from a colleague, or a financial update being broadcast to subscribers from a server. Ably uses channels to separate messages into different topics, so that clients only ever receive messages on the channels they are subscribed to.

### ChannelProvider <a id="step-2-channel-provider"/>

Now that you're connected to Ably, you can create and manage channels using the `ChannelProvider` component from the Ably Pub/Sub SDK. This component must be nested within the [`AblyProvider`](#prerequisites-setup-ably-provider) described above.

<Aside data-type='note'>
`ChannelProvider` is required when using feature hooks such as `useChannel()` or `usePresence()` exposed by the Ably Pub/Sub SDK.
</Aside>

Update your main `App` component to include the `ChannelProvider`:

<Code>
```react
// src/App.tsx

import { ChannelProvider } from 'ably/react';
import { ConnectionState } from './ConnectionState';

function App() {
  return (
    // Wrap components with ChannelProvider
    <ChannelProvider channelName='my-first-channel'>
      ...
    </ChannelProvider>
  );
}

export default App;
```
</Code>

### Subscribe to a channel <a id="step-2-subscribe-to-channel"/>

Use the `useChannel()` hook within the `ChannelProvider` component to subscribe to incoming messages on a channel. This hook also provides access to a `channel` instance and a `publish` method for sending messages.

In your project, create a new file called `src/Messages.tsx` and add new components called `Messages` and `MessageView`:

<Code>
```react
// src/Messages.tsx

import type { Message } from 'ably';
import { useChannel } from 'ably/react';
import { useState } from 'react';

function MessageView({ message }: { message: Message }) {
  // Displays an individual message
  const isMine = message.clientId === 'my-first-client';
  return (
    <p
      className={`py-1 px-2 shadow-sm ${
        isMine ? 'bg-green-100 text-gray-800' : 'bg-blue-50 text-gray-800'
      }`}
    >
      {message.data}
    </p>
  );
}

export function Messages() {
  const [messages, setMessages] = useState<Message[]>([]);

  // The useChannel hook subscribes to messages on the channel
  useChannel('my-first-channel', (message) => {
    setMessages((prevMessages) => [...prevMessages, message]);
  });

  return (
    <div className='flex flex-col w-full h-[600px] item-left rounded-lg overflow-hidden mx-auto font-sans'>
      <div className='flex-1 p-4 overflow-y-auto space-y-2'>
        {messages.map((msg: Message) => (
          <MessageView key={msg.id} message={msg} />
        ))}
      </div>
    </div>
  );
}
```
</Code>

Next, update your main `App` component in the `src/App.tsx` file to include the `Messages` component within the `ChannelProvider`:

<Code>
```react
// src/App.tsx

import { ChannelProvider } from 'ably/react';
import { ConnectionState } from './ConnectionState';
import { Messages } from './Messages';

function App() {
  return (
    <ChannelProvider channelName='my-first-channel'>
      <div className='flex flex-col w-[900px] h-full border-1 border-blue-500 rounded-lg overflow-hidden mx-auto font-sans'>
        ...
        <div className='flex flex-1 flex-row justify-evenly'>
          <div className='flex flex-col bg-white w-3/4 rounded-lg overflow-hidden mx-auto font-sans'>
            {/* Your Messages component should go here */}
            <Messages />
          </div>
        </div>
      </div>
    </ChannelProvider>
  );
}

export default App;
```
</Code>

You've successfully created a channel instance and set up a listener to receive messages. You can test this immediately by publishing messages using the Ably CLI:

<Code>
```shell
ably channels publish my-first-channel 'Hello from CLI!'
```
</Code>

### Publish a message <a id="step-2-publish-message"/>

You can publish messages in your React app using the `publish` method provided by the `useChannel()` hook.

Update your `src/Messages.tsx` file to include message publishing:

<Code>
```react
// src/Messages.tsx

// existing message imports and MessageView function

export function Messages() {
  const [messages, setMessages] = useState<Message[]>([]);
  const [inputValue, setInputValue] = useState('');

  // useChannel hook also provides a publish method
  const { publish } = useChannel('my-first-channel', (message) => {
    setMessages((prevMessages) => [...prevMessages, message]);
  });

  // Function to handle publishing messages
  const handlePublish = () => {
    if (!inputValue.trim()) return;
    publish('my-first-messages', inputValue.trim()).catch((err) =>
      console.error('Error publishing message', err)
    );
    setInputValue('');
  };

  return (
    <div className='flex flex-col w-full h-[600px] item-left rounded-lg overflow-hidden mx-auto font-sans'>
      <div className='flex-1 p-4 overflow-y-auto space-y-2'>
        {messages.map((msg: Message) => (
          <MessageView key={msg.id} message={msg} />
        ))}
      </div>
      <div className='flex items-center px-2 mt-auto mb-2'>
        <input
          type='text'
          placeholder='Type your message...'
          className='flex-1 p-2 border border-gray-400 rounded outline-none bg-white'
          value={inputValue}
          onChange={(e) => setInputValue(e.target.value)}
          onKeyDown={(event) => {
            if (event.key === 'Enter') {
              handlePublish();
            }
          }}
        />
        <button
          className='bg-blue-500 text-white px-4 ml-2 h-10 flex items-center justify-center rounded hover:bg-blue-600 transition-colors'
          onClick={handlePublish}
        >
          Publish
        </button>
      </div>
    </div>
  );
}
```
</Code>

Your application now supports publishing realtime messages! Type a message and click "Publish" to see it appear in your UI. Open another browser window to see clients interacting with each other in realtime or publish messages using the Ably CLI:

<Code>
```shell
ably channels publish my-first-channel 'Hello from CLI!'
```
</Code>

Messages from the CLI will appear in your UI in a different color to the ones you sent from the app.

## Step 3: Step 3: Join the presence set <a id="step-3"/>

Presence enables clients to be aware of one another if they are present on the same channel. You can then show clients who else is online, provide a custom status update for each, and notify the channel when someone goes offline.

Use the `usePresence()` and `usePresenceListener()` hooks provided by the Ably Pub/Sub SDK to interact with the presence feature in your React application. The `usePresence()` hook enables a client to join the presence set on a channel and update their presence status. The `usePresenceListener()` hook lets you subscribe to presence changes on a channel.

The `usePresenceListener()` hook also returns an object containing the `presenceData` array, which holds current presence data on the channel.

Create a new file called `src/PresenceStatus.tsx` with the following content:

<Code>
```react
// src/PresenceStatus.tsx

// 'ably/react' exports hooks for working with presence on a channel
import { usePresence, usePresenceListener } from 'ably/react';

export function PresenceStatus() {
  // Enter the current client into the presence set with an optional status
  usePresence('my-first-channel', { status: "I'm here!" });

  // Subscribe to presence updates on the channel
  const { presenceData } = usePresenceListener('my-first-channel');

  return (
    <div className='flex flex-col bg-white w-full h-full px-4 py-2'>
      <strong className='text-green-700 mr-4 text-center border-b border-gray-900'>
        Present: {presenceData.length}
      </strong>

      <div className='flex-1 flex-col flex flex-nowrap items-start gap-4 overflow-x-auto'>
        {presenceData.map((member, idx) => (
          <div key={idx} className='flex items-center gap-1'>
            <span className='inline-block w-2 h-2 rounded-full bg-green-500' />
            <span className='text-gray-800'>
              {member.clientId}
              {member.data?.status ? ` (${member.data.status})` : ''}
            </span>
          </div>
        ))}
      </div>
    </div>
  );
}
```
</Code>

Add the `PresenceStatus` component to your main `App` component in `src/App.tsx` as follows:

<Code>
```react
// src/App.tsx

// Existing imports
import { PresenceStatus } from './PresenceStatus';

function App() {
  return (
    <ChannelProvider channelName='my-first-channel'>
      <div className='flex flex-col w-[900px] h-full border-1 border-blue-500 rounded-lg overflow-hidden mx-auto font-sans'>
        <div className='flex flex-row w-full rounded-lg overflow-hidden mx-auto font-sans'>
          <div className='flex-1 bg-gray-100 text-center p-4'>
            <h2 className='text-lg font-semibold text-blue-500'>
              Ably Pub/Sub React
            </h2>
            <ConnectionState />
          </div>
        </div>
        <div className='flex flex-1 flex-row justify-evenly'>
          <div className='flex flex-col w-1/4 border-r-1 border-blue-500 overflow-hidden mx-auto font-sans'>
            <div className='flex-1 overflow-y-auto'>
              {/* Your PresenceStatus component should go here */}
              <PresenceStatus />
            </div>
          </div>

          <div className='flex flex-col bg-white w-3/4 rounded-lg overflow-hidden mx-auto font-sans'>
            <Messages />
          </div>
        </div>
      </div>
    </ChannelProvider>
  );
}

export default App;
```
</Code>

The application will now display a list of clients currently present on the channel. The `usePresence()` hook enters your client into the channel's presence set with an optional status, while the `usePresenceListener()` hook subscribes to presence updates. Your current client ID should appear in the list of online users.

You can have another client join the presence set using the Ably CLI:

<Code>
```shell
ably channels presence enter my-first-channel --client-id "my-cli" --data '{"status":"From CLI"}'
```
</Code>

## Step 4: Retrieve message history <a id="step-4"/>

You can retrieve previously sent messages using the history feature. Ably stores all messages for 2 minutes by default in the event a client experiences network connectivity issues. This can be extended for longer if required.

Although the Ably Pub/Sub SDK does not provide a specific hook for retrieving message history, you can use the `useChannel()` hook to get a [`RealtimeChannel`](https://ably.com/docs/sdk/js/v2.0/interfaces/ably.RealtimeChannel.html) instance and then call its [`history()`](https://ably.com/docs/sdk/js/v2.0/interfaces/ably.RealtimeChannel.html#history) method to retrieve messages recently published to the channel.

Update your `src/Messages.tsx` file to include the new `useEffect` within your existing `Messages` component:

<Code>
```react
// src/Messages.tsx

// Existing imports
import { useEffect, useState } from 'react';

// MessageView function remains unchanged

export function Messages() {
  // useStates, useChannel, and handlePublish function remain unchanged

  useEffect(() => {
    async function loadHistory() {
      try {
        // Retrieve the last 5 messages from history
        const history = await channel.history({ limit: 5 });
        // History responses are returned in reverse chronological order (newest first)
        // Reverse the array to show the latest messages at the bottom in the UI
        const messagesFromHistory = history.items.reverse();
        // Update the state with retrieved messages
        setMessages(messagesFromHistory);
      } catch (error) {
        console.error('Error loading message history:', error);
      }
    }

    loadHistory();
  }, [channel]);

  // Return remains unchanged
}
```
</Code>

Test this feature with the following steps:

1. Publish several messages using your application UI, or send messages from another client using the Ably CLI:

<Code>
```shell
ably channels publish --count 5 my-first-channel "Message number {{.Count}}"
```
</Code>

2. Refresh the page. This will cause the `Messages` component to mount again and call the `channel.history()` method.
3. You should see the last 5 messages displayed in your UI, ordered from oldest to newest at the bottom:

<Code>
```text
Message number 1
Message number 2
Message number 3
Message number 4
Message number 5
```
<Code>

## Step 5: Close the connection <a id="step-5"/>

Connections are automatically closed approximately two minutes after the last channel is detached. However, explicitly closing connections when they're no longer needed is good practice to help save costs and clean up listeners.

<Code>
```react
const realtimeClient = useAbly();

const handleDisconnect = () => {
  realtimeClient.connection.close();
};

// Call handleDisconnect when needed
```
</Code>

This ensures the connection is closed when your component unmounts, freeing resources and removing listeners.

## Next steps <a id="next-steps"/>

Continue to explore the Ably Pub/Sub documentation with React as the selected language:

Read more about the concepts covered in this guide:

* Revisit the basics of [Pub/Sub](/docs/pub-sub)
* Explore more [advanced](/docs/pub-sub/advanced) Pub/Sub concepts
* Understand realtime [connections](/docs/connect) to Ably
* Read more about how to use [presence](/docs/presence-occupancy/presence) in your apps
* Fetch message [history](/docs/storage-history/history) in your apps

You can also explore the [Ably CLI](https://www.npmjs.com/package/@ably/cli) further, or visit the Pub/Sub [API references](/docs/api/realtime-sdk).
