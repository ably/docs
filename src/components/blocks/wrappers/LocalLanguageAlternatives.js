import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { DEFAULT_LANGUAGE } from '../../../../data/createPages/constants';
import languageLabels from '../../../maps/language';
import MenuItemButton, { SelectedMenuItemButton } from '../../Menu/MenuItem/MenuItemButton';
import Html from '../Html';
import { ChildPropTypes } from '../../../react-utilities';
import LanguageNavigation from '../../Menu/LanguageNavigation';

const LocalLanguageAlternatives = ({ language, languages, data, children }) => {
  const [selected, setSelected] = useState(children);
  const [selectedLanguage, setSelectedLanguage] = useState(languages[0]);

  const onClick = ({ target: { value } }) => {
    setSelected(<Html data={data[value]} />);
    setSelectedLanguage(value);
  };

  const label = language === DEFAULT_LANGUAGE ? `SELECT` : `No ${languageLabels[language]} example exists, select:`;
  const languageItems = languages.map((lang) => ({
    Component: lang === selectedLanguage ? SelectedMenuItemButton : MenuItemButton,
    props: {
      onClick,
      value: lang,
    },
    content: languageLabels[lang] ?? '',
  }));

  return (
    <>
      <LanguageNavigation label={label} items={languageItems} />
      {selected}
    </>
  );
};

LocalLanguageAlternatives.propTypes = {
  language: PropTypes.string,
  languages: PropTypes.array,
  data: PropTypes.object,
  children: ChildPropTypes,
};

export default LocalLanguageAlternatives;
