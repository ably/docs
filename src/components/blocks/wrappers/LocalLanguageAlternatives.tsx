import React, { useContext, useState } from 'react';
import { DEFAULT_LANGUAGE } from '../../../../data/createPages/constants';
import languageLabels from '../../../maps/language';
import MenuItemButton, { SelectedMenuItemButton } from '../../Menu/MenuItem/MenuItemButton';
import Html from '../Html';
import LanguageNavigation from '../../Menu/LanguageNavigation';
import { PageLanguagesContext } from '../../../contexts/page-language-context';
import LanguageButton from '../../Button/LanguageButton';

const LocalLanguageAlternatives = ({
  language,
  languages,
  data,
  children,
}: {
  language: string;
  languages: string[];
  data: Record<string, string | any[] | null | undefined> | undefined;
  children: React.ReactNode;
}) => {
  const pageLanguages = useContext(PageLanguagesContext);

  const [selected, setSelected] = useState(children);
  const [selectedLanguage, setSelectedLanguage] = useState(languages[0]);

  const onClick = ({ target: { value } }: { target: { value: string } }) => {
    setSelected(<Html data={data ? data[value] : ''} />);
    setSelectedLanguage(value);
  };

  const label = [...languages, DEFAULT_LANGUAGE].includes(language)
    ? `SELECT`
    : `No ${languageLabels[language] ?? language} example exists, select:`;
  const languageItems = languages.map((lang) => {
    // Site navigation button
    if (pageLanguages.includes(lang)) {
      return {
        Component: LanguageButton,
        props: { language: lang },
        content: languageLabels[lang] ?? lang,
      };
    }
    // Local button, if global language option doesn't exist
    return {
      Component: lang === selectedLanguage ? SelectedMenuItemButton : MenuItemButton,
      props: {
        language: '',
        onClick,
        value: lang,
      },
      content: languageLabels[lang] ?? lang,
    };
  });

  return (
    <>
      <LanguageNavigation label={label} items={languageItems} />
      {selected}
    </>
  );
};

export default LocalLanguageAlternatives;
