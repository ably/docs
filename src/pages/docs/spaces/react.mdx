---
title: React Hooks
meta_description: "Incorporate Spaces into your React application with idiomatic and user-friendly React Hooks."
---

Incorporate Spaces into your React application with idiomatic and user-friendly React Hooks.

This package enables you to:

* Interact with Ably [Spaces](/docs/spaces) using a React Hook
* Subscribe to [events](/docs/spaces/avatar#events) in a space
* Retrieve the [membership](/docs/spaces/space) of a space
* Set the [location](/docs/spaces/locations) of space members
* Acquire [locks](/docs/spaces/locking) on components within a space
* Set the position of [members' cursors](/docs/spaces/cursors) in a space

The following hooks are available:

| Hook | Description |
| ---- | ----------- |
| [`useSpace`](#useSpace) | The `useSpace` hook lets you subscribe to the current Space, receive Space state events, and get the current Space instance. |
| [`useMembers`](#useMembers) | The `useMembers` hook is useful in building avatar stacks. Using `useMembers`, you can retrieve space members. |
| [`useLocations`](#useLocations) | The `useLocations` hook lets you subscribe to location events. Location events are emitted whenever a member changes location. |
| [`useLocks`](#useLocks) | The `useLocks` hook lets you subscribe to lock events by registering a listener. Lock events are emitted whenever the lock state transitions into `locked` or `unlocked`. |
| [`useLock`](#useLock) | The `useLock` hook returns the status of a lock and, if it has been acquired, the member holding the lock. |
| [`useCursors`](#useCursors) | The `useCursors` hook allows you to track a member's pointer position updates across an application. |

Spaces hooks are client-side oriented. If you employ server-side rendering, ensure components using these hooks render only on the client-side.

## Install <a id="install"/>

<Code>
```shell
npm install ably @ably/spaces
```
</Code>

<Aside data-type='note'>
React version 16.8.0 or above is required.
</Aside>

## Authenticate <a id="authenticate"/>

An [API key](/docs/auth#api-keys) is required to authenticate with Ably. API keys are used either to authenticate directly with Ably using [basic authentication](/docs/auth/basic), or to generate tokens for untrusted clients using [token authentication](/docs/auth/token).

[Sign up](https://ably.com/sign-up) to Ably to create an API key in the [dashboard](https://ably.com/dashboard) or use the [Control API](/docs/platform/account/control-api) to create an API programmatically.

<Aside data-type='important'>
The examples use [basic authentication](/docs/auth/basic) to demonstrate usage for convenience. In your own applications, basic authentication should never be used on the client-side as it exposes your Ably API key. Instead use [token authentication](/docs/auth/token).
</Aside>

## Usage <a id="usage"/>

### Setting up the Spaces Provider <a id="spaces-provider"/>

Use the `SpacesProvider` component to connect to Ably. The `SpacesProvider` should wrap every component that needs to access Spaces.

<Code>
```react
import { Realtime } from "ably";
import Spaces from "@ably/spaces";
import { SpacesProvider, SpaceProvider } from "@ably/spaces/react";

const ably = new Realtime({ key: "{{API_KEY}}", clientId: 'clemons' });
const spaces = new Spaces(ably);

root.render(
  <SpacesProvider client={spaces}>
    <SpaceProvider name="my-space">
      <App />
    </SpaceProvider>
  </SpacesProvider>
)
```
</Code>

### useSpace <a id="useSpace"/>

The `useSpace` hook enables you to subscribe to the current [Space](/docs/spaces/space), receive Space state events, and get the current Space instance.

<Code>
```react
const { space } = useSpace((update) => {
  console.log(update);
});
```
</Code>

### useMembers <a id="useMembers"/>

`useMembers` is used to build [avatar stacks](/docs/spaces/avatar). It retrieves members of the space, including members that have recently left the space, but have not yet been removed.

<Code>
```react
const { self, others, members } = useMembers();
```
</Code>

- `self` - a memberâ€™s own member object
- `others` - an array of member objects for all members other than the member themselves
- `members` - an array of all member objects, including the member themselves

`useMembers` also enables you to subscribe to members entering, leaving, and being removed from the Space (after a timeout), as well as when a member updates their [profile information](/docs/spaces/space#update-profile).

<Code>
```react
// Subscribe to all member events in a space
useMembers((memberUpdate) => {
  console.log(memberUpdate);
});

// Subscribe to member enter events only
useMembers('enter', (memberJoined) => {
  console.log(memberJoined);
});

// Subscribe to member leave events only
useMembers('leave', (memberLeft) => {
  console.log(memberLeft);
});

// Subscribe to member remove events only
useMembers('remove', (memberRemoved) => {
  console.log(memberRemoved);
});

// Subscribe to profile updates on members only
useMembers('updateProfile', (memberProfileUpdated) => {
  console.log(memberProfileUpdated);
});

// Subscribe to all updates to members
useMembers('update', (memberUpdate) => {
  console.log(memberUpdate);
});
```
</Code>

### useLocations <a id="useLocations"/>

`useLocations` enables you to subscribe to [location](/docs/spaces/locations) events. Location events are emitted whenever a member changes location.

<Code>
```react
useLocations((locationUpdate) => {
  console.log(locationUpdate);
});
```
</Code>

`useLocations` also enables you to update the current member location using the `update` method provided by the hook. For example:

<Code>
```react
const { update } = useLocations((locationUpdate) => {
  console.log(locationUpdate);
});
```
</Code>

### useLocks <a id="useLocks"/>

`useLocks` enables you to subscribe to [lock](/docs/spaces/locking) events by registering a listener. Lock events are emitted whenever a lock transitions into the `locked` or `unlocked` state.

<Code>
```react
useLocks((lockUpdate) => {
  console.log(lockUpdate);
});
```
</Code>

### useLock <a id="useLock"/>

`useLock` returns the status of a [lock](/docs/spaces/locking) and, if the lock has been acquired, the member holding that lock.

<Code>
```react
const { status, member } = useLock('my-lock');
```
</Code>

### useCursors <a id="useCursors"/>

`useCursors` enables you to track a member's [cursor](/docs/spaces/cursors) position and provide a view of all members' cursors within a space. For example:

<Code>
```react
// Subscribe to events published on "mousemove" by all members
const { set } =  useCursors((cursorUpdate) => {
  console.log(cursorUpdate);
});

useEffect(() => {
  // Publish a your cursor position on "mousemove" including optional data
  const eventListener = ({ clientX, clientY }) => {
    set({ position: { x: clientX, y: clientY }, data: { color: 'red' } });
  }

  window.addEventListener('mousemove', eventListener);

  return () => {
    window.removeEventListener('mousemove', eventListener);
  };
});
```
</Code>

If you provide `{ returnCursors: true }` as an option you can retrieve active members' cursors:

<Code>
```react
const { cursors } =  useCursors((cursorUpdate) => {
  console.log(cursorUpdate);
}, { returnCursors: true });
```
</Code>

### Error handling <a id="error-handling"/>

[`useSpace`](#useSpace), [`useMembers`](#useMembers), [`useLocks`](#useLocks), and [`useCursors`](#useCursors) return [connection](/docs/connect) and [channel](/docs/channels) errors you may encounter, so that you can handle them within your components. This may include when a client doesn't have permission to attach to a channel, or if it loses its connection to Ably.

<Code>
```react
const { connectionError, channelError } = useMembers();

if (connectionError) {
  // TODO: handle connection errors
} else if (channelError) {
  // TODO: handle channel errors
} else {
  return <SpacesPoweredComponent />
}
```
</Code>
