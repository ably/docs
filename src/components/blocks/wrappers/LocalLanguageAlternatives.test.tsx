import React from 'react';
import { render, screen } from '@testing-library/react';

import LocalLanguageAlternatives from './LocalLanguageAlternatives';

const props = {
  data: {
    javascript: [
      {
        data: [
          {
            data: 'This is JavaScript.',
            type: 'text',
            name: 'text',
          },
        ],
        type: 'tag',
        name: 'code',
        attribs: {
          lang: 'javascript',
        },
      },
    ],
    ruby: [
      {
        data: [
          {
            data: 'This is Ruby.',
            type: 'text',
            name: 'text',
          },
        ],
        type: 'tag',
        name: 'code',
        attribs: {
          lang: 'ruby',
        },
      },
    ],
    csharp: [
      {
        data: [
          {
            data: 'This is C#.',
            type: 'text',
            name: 'text',
          },
        ],
        type: 'tag',
        name: 'code',
        attribs: {
          lang: 'csharp',
        },
      },
    ],
  },
  languages: ['javascript', 'ruby', 'csharp'],
  localChangeOnly: false,
  initialData: [
    {
      data: [{ data: "await ably.close()\nprint('Closed the connection to Ably.')", type: 'text', name: 'text' }],
      type: 'tag',
      name: 'code',
      attribs: { lang: 'python' },
    },
  ],
};

describe('<LocalLanguageAlternatives />', () => {
  it('renders correctly', () => {
    const { container } = render(<LocalLanguageAlternatives {...props} />);
    expect(container).toMatchSnapshot();
    // expect(screen.getByTestId('menu')).toBeInTheDocument();
  });

  it("doesn't render the menu bar when only one language", () => {
    render(<LocalLanguageAlternatives {...props} languages={['javascript', '']} />);
    expect(screen.queryByRole('menu')).not.toBeInTheDocument();
  });
});
