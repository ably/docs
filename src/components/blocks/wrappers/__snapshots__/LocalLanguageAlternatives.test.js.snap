// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalLanguageAlternatives displays statically as expected Snapshot of LocalLanguageAlternatives is the same 1`] = `
Array [
  <menu
    className="sc-fEOsli sc-bjUoiL sc-idiyUo kSqfms kaHjhA JUXrX"
  >
    <li
      className="c-menu__item c-menu__label"
    >
      Select:
    </li>
    <li
      className="c-menu__item"
    >
      <button
        className="sc-bczRLJ sc-gsnTZi sc-dkzDqf iBbGtg hPThGj giflia"
        onClick={[Function]}
        value="javascript"
      >
        JS
      </button>
    </li>
    <li
      className="c-menu__item"
    >
      <button
        className="sc-bczRLJ sc-gsnTZi iBbGtg hPThGj"
        onClick={[Function]}
        value="ruby"
      >
        Ruby
      </button>
    </li>
    <li
      className="c-menu__item"
    >
      <button
        className="sc-bczRLJ sc-gsnTZi iBbGtg hPThGj"
        onClick={[Function]}
        value="csharp"
      >
        C#
      </button>
    </li>
  </menu>,
  "This is default content.",
]
`;
